compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -2648,6 +2648,78 @@ declare_lint! {
     };
 }
 
+declare_lint! {
+    /// The `fuzzy_provenance_casts` lint detects an `as` cast between an integer
+    /// and a pointer.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn main() {
+    ///     let dangling = 16_usize as *const u8;
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// This lint is part of the strict provenance effort, see [issue #95228].
+    /// Casting an integer to a pointer is considered bad style, as a pointer
+    /// contains, besides the *address* also a *provenance*, indicating what
+    /// memory the pointer is allowed to read/write. Casting an integer, which
+    /// doesn't have provenance, to a pointer requires the compiler to assign
+    /// (guess) provenance. The compiler assigns "all valid" provenance, but
+    /// this penalizes the optimiser and is not well suited for dynamic
+    /// analysis/dynamic program verification (e.g. `miri` or CHERI platforms).
+    ///
+    /// It is much better to use `ptr::with_addr` instead to specify the
+    /// provenance you want.
+    ///
+    /// This might (or might not) become a [future-incompatible] lint or hard
+    /// error in a future rust edition. See [issue #95228] for more details.
+    ///
+    /// [future-incompatible]: ../index.md#future-incompatible-lints
+    /// [issue #95228]: https://github.com/rust-lang/rust/issues/95228
+    pub FUZZY_PROVENANCE_CASTS,
+    Allow,
+    "A fuzzy integer to pointer cast is used",
+    @feature_gate = sym::strict_provenance_lints;
```

Comment:
Would it make sense to use the same feature gate (name) as the library functions, `#![feature(strict_provenance)]`? Would that cause problems?

Reply:
Yes, that probably make sense.

Reply:
I did this (in a separate commit so it is easy to revert) but I am now in doubt of whether these features shoudn't remain separate. It forces all documentation of the feature gate to refer to  #95228 instead of #95488. And it is kind of ugly and possibly confusing that the same name is used for both a library feature and a compiler feature.

Reply:
> Merging them forces all documentation of the feature gate to refer to https://github.com/rust-lang/rust/issues/95228 instead of https://github.com/rust-lang/rust/issues/95488.

I think this is the right thing to do anyway, because #95488 isn't a tracking issue.

> And it is kind of ugly and possibly confusing that the same name is used for both a library feature and a compiler feature.

On the other hand, the only concrete advantage of having separate feature gates is to allow both features to be stabilized separately, but I don't think it would make sense here. I also think that it would be more confusing to enable `#![feature(strict_provenance_lints)]` and not be able to use the appropriate APIs without another feature gate.

Reply:
True. Are there other features which are both a library and a compiler feature?

Reply:
I have no idea :sweat_smile: